lib/Sema/CSStep.cpp

```diff
@@ -329,7 +329,8 @@ StepResult ComponentStep::take(bool prevFailed) {
   auto *disjunction = CS.selectDisjunction();
   auto bestBindings = CS.determineBestBindings();
 
-  if (bestBindings && (!disjunction || (!bestBindings->InvolvesTypeVariables &&
+  if (bestBindings && (!disjunction || (!bestBindings->IsHole &&
```

Comment:
I thought that idea was to continue attempting type variables until there are no more bindings to attempt. If we disallow bindings which represent holes I don't think we'll be able to solve situations like this:

```swift
struct S {
  func foo<T>(_ a: T, _ b: Int) -> Void {}
  func foo(_ a: String) -> Void {}

  func test() -> Void {
    foo(.hello, 42)
  }
}
```

Where `T` of `foo` is going to be a base type for `.hello` which means that we probably have to attempt `T := UnresolvedType` to be able to propagate holes to the result type variable of the `unresolved member` reference and the rest of the constraint system.

Reply:
This idea is to not choose a hole binding if there is a disjunction. So, we will always pick an available disjunction over a binding for a hole. If there is no disjunction, we will attempt the hole binding. The logic is kind of hard to read though :/ 

Your test case diagnoses:
```
/Users/hborla/Desktop/test.swift:69:5: error: generic parameter 'T' could not be inferred
    foo(.hello, 42)
    ^
/Users/hborla/Desktop/test.swift:65:8: note: in call to function 'foo'
  func foo<T>(_ a: T, _ b: Int) -> Void {}
       ^
```

Reply:
In other words, we only choose a hole binding if there are no other bindings to try _and_ no disjunctions to try

Reply:
Ah right, sorry I misread `!disjunction` bit.

I gets correctly diagnosed even right now but it's because of CSDiag trying re-typecheck it piecemeal. What I'm wondering is - with your changes does it get diagnosed by a fix without any CSDiag involvement?

Reply:
Sorry, yes, I meant that I just tested it with my change and it is diagnosed properly without going through CSDiag.

Reply:
Great! Thank you for conforming!